src/H5Dmpio.c

```diff
@@ -2954,464 +3398,1569 @@ H5D__chunk_redistribute_shared_chunks(const H5D_io_info_t *io_info, const H5D_ty
+ * Function:    H5D__mpio_get_chunk_fspace_info_types
+ *
+ * Purpose:     Constructs MPI derived datatypes for communicating the
+ *              inner `fspace_info` structure of a
+ *              H5D_filtered_collective_io_info_t structure.
+ *
+ *              The datatype returned through `struct_type` has an extent
+ *              equal to the size of the `fspace_info` structure and is
+ *              suitable for receiving one or more `fspace_info` structures
+ *              into a contiguous array.
+ *
+ *              The datatype returned through `resized_type` has a lower
+ *              bound equal to the offset of the `fspace_info` structure
+ *              inside the encompassing H5D_filtered_collective_io_info_t
+ *              structure and an extent equal to the size of the
+ *              encompassing structure. This makes it suitable for
+ *              sending an array of H5D_filtered_collective_io_info_t
+ *              structures, while extracting out just the nested
+ *              `fspace_info` structure during communication.
+ *
+ * Return:      Non-negative on success/Negative on failure
+ *
+ *-------------------------------------------------------------------------
+ */
+static herr_t
+H5D__mpio_get_chunk_fspace_info_types(MPI_Datatype *struct_type, hbool_t *struct_type_derived,
+                                      MPI_Datatype *resized_type, hbool_t *resized_type_derived)
+{
+    MPI_Datatype hindexed_type         = MPI_DATATYPE_NULL;
+    hbool_t      hindexed_type_derived = FALSE;
+    MPI_Datatype inner_type            = MPI_DATATYPE_NULL;
+    hbool_t      inner_type_derived    = FALSE;
+    int          mpi_code;
+    herr_t       ret_value = SUCCEED;
+
+    FUNC_ENTER_STATIC
+
+    HDassert(struct_type);
+    HDassert(struct_type_derived);
+    HDassert(resized_type);
+    HDassert(resized_type_derived);
+
+    *struct_type_derived  = FALSE;
+    *resized_type_derived = FALSE;
+
+    /* Create struct type for the inner H5F_block_t structure */
+    if (H5F_mpi_get_file_block_type(FALSE, &inner_type, &inner_type_derived) < 0)
+        HGOTO_ERROR(H5E_DATASET, H5E_CANTGET, FAIL, "can't create derived type for chunk file description")
+
+    /* Create contiguous type for `fspace_info` structure */
+    if (MPI_SUCCESS != (mpi_code = MPI_Type_contiguous(2, inner_type, struct_type)))
+        HMPI_GOTO_ERROR(FAIL, "MPI_Type_contiguous failed", mpi_code)
+    *struct_type_derived = TRUE;
+
+    if (MPI_SUCCESS != (mpi_code = MPI_Type_commit(struct_type)))
+        HMPI_GOTO_ERROR(FAIL, "MPI_Type_commit failed", mpi_code)
+
+    /* Created hindexed block type with an offset equal to the `fspace_info`
+     * structure's offset in the encompassing H5D_filtered_collective_io_info_t
+     * structure. Then, resize the type with an extent equal to the size of the
+     * encompassing structure to make it suitable for sending arrays of them.
+     */
+    {
```

Comment:
was there a reason for making a separate block?

Reply:
In this case, it was just to have a smaller scope for the 'extent' variable that we throw away after the call to MPI_Type_get_extent. To create the resized type, I only needed the lower bound ('lb') for the original type and not the extent, but I believe MPI_Type_get_extent requires you to provide variables for both 'lb' and 'extent'.